### ISSUE TITLE
Best way to strip script tags and strip onClick, onBlur and other JS events?  

### ISSUE THREAD

**Issue by dkwin** [markdown-it/markdown-it#305]
I'm overriding render.rules.html_inline and searching for onClick, onXYZ events and removing them.  Also, just escaping <script> tags entirely.

I'm wondering if there is a better way and more efficient way to do this?

**Comment by puzrin** [comment]
Not sure i understand your question. Could you rephrase?

**Comment by dkwin** [comment]
Sorry for not clarifying, I'm trying to allow html tags within the markup editor but I want to guard against XSS.  

for example
`<i onClick ="alert ('hacked')" onBlur ="alert ('hacked')"> testing onClick JS </i>`

or
`<script> alert ("hacked since this script tag runs </script>`

Here is the code that I have
`var previousInlineHTML = mdParser.renderer.rules.html_inline;

mdParser.renderer.rules.html_inline = (tokens, idx, options, env, self) => {
    tokens[idx].content = escapePotentialJSInAttributse (tokens[idx].content);
    if (tokens[idx].content.startsWith("<script")){
        tokens[idx].content = mdParser.utils.escapeHtml (tokens[idx].content);
    }
    return previousInlineHTML (tokens, idx, options, env, self);
};

const escapePotentialJSInAttributes= (string) => {
    return string.replace(/ on\w+="[^"]*"/g, '');
}`


This isn't optimal but I was wondering if intercepting the tags at the rule level was the right approach or if there is a better way of handling this to protect against XSS.

**Comment by puzrin** [comment]
That's impossible to do correct. You should disable html or use external validation package.

**Comment by dkwin** [comment]
@puzrin 

Is it possible to limit to just some tags like <i> or <b> and disable the rest?  Or would I have to intercept every html_inline token?

How could I even remove tokens from being rendered?

**Comment by puzrin** [comment]
As i said, you can not process html_inline tokens and guarantee correct output. Use external sanitizer for that. This package is not designed for such tasks. It can produce correct output only if html is disabled.

**Comment by dkwin** [comment]
@puzrin 

Thanks for the reply

Your recommendation is just to scrub the actual text/string before it even goes to markdown-it's render method?

I'm curious as to how XSS is usually handled when allowing html to be used with markdown-it?  

Github's markdown editor actually allows some html tags but does not allow onClick, onBlur, etc.

**Comment by puzrin** [comment]
If you need html tags in markdown, pass output after markdown parser to external sanitizer to make it safe. See https://github.com/npm/marky-markdown sources for example.

**Comment by rlidwka** [comment]
@dkwin , use a sanitizer.

Here is an example which closely matches github editor behavior: https://github.com/rlidwka/render-readme/blob/6b74013ce0388f962a27ae487bc55a5f25b2b9f1/lib/sanitize.js

**Comment by dkwin** [comment]
@puzrin 

thanks for the info.  Much appreciated.
